internal/node/hybrid/ip_validator.go

```
@@ -0,0 +1,221 @@
+	}
+	if nodeIP.IsMulticast() {
+		return fmt.Errorf("nodeIP can't be a multicast address")
+	}
+	if nodeIP.IsLinkLocalUnicast() {
+		return fmt.Errorf("nodeIP can't be a link-local unicast address")
+	}
+	if nodeIP.IsUnspecified() {
+		return fmt.Errorf("nodeIP can't be an all zeros address")
+	}
+
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return err
+	}
+	for _, addr := range addrs {
+		var ip net.IP
+		switch v := addr.(type) {
+		case *net.IPNet:
+			ip = v.IP
+		case *net.IPAddr:
+			ip = v.IP
+		}
+		if ip != nil && ip.Equal(nodeIP) {
+			return nil
+		}
+	}
+	return fmt.Errorf("node IP: %q not found in the host's network interfaces", nodeIP.String())
+}
+
+// getNodeIP determines the node's IP address based on kubelet configuration and system information.
+func getNodeIP(kubeletArgs []string, iamNodeName string) (net.IP, error) {
+	// Follows algorithm used by kubelet to assign nodeIP
+	// Implementation adapted for hybrid nodes
+	// 1) Use nodeIP if set (and not "0.0.0.0"/"::")
+	// 2) If the user has specified an IP to HostnameOverride, use it (not allowed for hybrid nodes)
+	// 3) Lookup the IP from node name by DNS
+	// 4) Try to get the IP from the network interface used as default gateway
+	// Source: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L206
+
+	if err := ensureNoHostnameOverride(kubeletArgs); err != nil {
+		return nil, err
+	}
+
+	nodeIP, err := extractNodeIPFromFlags(kubeletArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	var ipAddr net.IP
+
+	nodeIPSpecified := nodeIP != nil && nodeIP.To4() != nil && !nodeIP.IsUnspecified()
+
+	if nodeIPSpecified {
+		ipAddr = nodeIP
+	} else {
+		// If using SSM, the node name will be set at initialization to the SSM instance ID,
+		// so it won't resolve to anything via DNS, hence we're only checking in the case of IAM-RA
+		if iamNodeName != "" {
+			var addrs []net.IP
+
+			addrs, _ = net.LookupIP(iamNodeName)
```

any reason to ignore the error?

---

this part is from the kubelet code for the way it sets the ip, and we just want to replicate it

---

What does this mean, that we ignore the error because the kubelet ignores the error?

If so, then i think we want to understand why the kuebelet ignores it, not just blindly copy it